docs: Add architecture overview (#2816)
oasislabs/oasis-core: docs/consensus/registry.md

# Registry

> The registry service is responsible for managing an entity/node/runtime public
> key and metadata registry.

Reviewer: ```suggestion
The registry service is responsible for managing a registry of entity/node/runtime
public keys and metadata.
```

The service interface definition lives in [`go/registry/api`]. It defines the
supported queries and transactions. For more information you can also check out
the [consensus service API documentation].

<!-- markdownlint-disable line-length -->
[`go/registry/api`]: ../../go/registry/api
[consensus service API documentation]: https://pkg.go.dev/github.com/oasislabs/oasis-core/go/registry/api?tab=doc
<!-- markdownlint-enable line-length -->

## Methods

### Register Entity

> Entity registration enables a new entity to be created. A new register entity

Reviewer: We should define somewhere (probably not here) what an "entity" is. This text should probably link to that definition.

Reviewer: For reference, this is what we have in the operator docs: https://docs.oasis.dev/operators/architecture-overview.html#entities-and-key-management. It's a bit high-level/casual, matching the overall tone of those docs.

Author: Added.

* Node identity key.
* Consensus key.
* Committee TLS certificate key.
* Committee P2P key.

Registering a node may require sufficient stake in the owning entity's
[escrow account].

<!-- markdownlint-disable line-length -->
[`NewRegisterNodeTx`]: https://pkg.go.dev/github.com/oasislabs/oasis-core/go/registry/api?tab=doc#NewRegisterNodeTx
[`MultiSignedNode`]: https://pkg.go.dev/github.com/oasislabs/oasis-core/go/common/node?tab=doc#MultiSignedNode
[`Node`]: https://pkg.go.dev/github.com/oasislabs/oasis-core/go/common/node?tab=doc#Node
[multi-signed envelope]: ../crypto.md#multi-signed-envelope
<!-- markdownlint-enable line-length -->

### Unfreeze Node

Node unfreezing enables a previously frozen (e.g., due to slashing) node to be
thawed so it can again be eligible for committee elections. A new unfreeze node
transaction can be generated using [`NewUnfreezeNodeTx`].

**Method name:**

```
registry.UnfreezeNode
```

**Body:**

```golang
type UnfreezeNode struct {
    NodeID signature.PublicKey `json:"node_id"`
}
```

**Fields:**

* `node_id` specifies the node identifier of the node to thaw.

The transaction signer MUST be the entity key that owns the node.

> Thawing a node requires that the node's freeze period has already passed.

Reviewer: Might be useful to note that the duration of the freeze period for a given event (e.g., double signing) is a consensus parameter.

<!-- markdownlint-disable line-length -->
[`NewUnfreezeNodeTx`]: https://pkg.go.dev/github.com/oasislabs/oasis-core/go/registry/api?tab=doc#NewUnfreezeNodeTx
<!-- markdownlint-enable line-length -->

### Register Runtime

Runtime registration enables a new runtime to be created. A new register
runtime transaction can be generated using [`NewRegisterRuntimeTx`].

**Method name:**

```
registry.RegisterRuntime
```

The body of a register runtime transaction must be a [`SignedRuntime`]
structure, which is a [signed envelope] containing a [`Runtime`] descriptor.
The signer of the transaction MUST be the owning entity key.

Registering a runtime may require sufficient stake in the owning entity's

> [escrow account].

Reviewer: Might be useful to point out the consensus parameter that defines this threshold.
